### PR TITLE
Fallback to default visualization based on type, when no viz has been explicitly set

### LINKED ISSUE #8386: Default visualization
Before the user picks a concrete visualization, the chosen vis should be "default" - the exact kind should be picked according to the node output type (consult the old GUI for details - most prominently, the `Table` type displays `Table` visualization).

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
- Closes #8386
  - Attempts to execute `<expr>.default_visualization` to query the engine for the *correct* fallback type
  - If that is not possible, falls back to checking `inputType` - first for an exact match, then falling back to `Any` (i.e. the text/JSON visualization)
  - Does not decide fallback based on the shape of the returned JSON

### Important Notes
Contains pretty significant refactors of `VisualizationMetadata` to allow it to be unset.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] ~~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

### CODE REVIEW

File: app/gui2/shared/yjsModel.ts [hunk 1]
```diff
@@ -28,7 +29,13 @@ export function visMetadataEquals(
...
   return (
     (a == null && b == null) ||
-    (a != null && b != null && a.visible === b.visible && visIdentifierEquals(a, b))
+    (a != null &&
+      b != null &&
+      a.visible === b.visible &&
+      ((a.identifier == null && b.identifier == null) ||
+        (a.identifier != null &&
+          b.identifier != null &&
+          visIdentifierEquals(a.identifier, b.identifier))))
```
Comment (farmaazon) [inline app/gui2/shared/yjsModel.ts]: Please split this expression, maybe to some `if`. Or perhaps `visIdentifierEquals` should accept nulls?
You can also use `compareOpt`.

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue [hunk 2]
```diff
@@ -36,6 +39,62 @@ const emit = defineEmits<{
...
+const graphStore = useGraphStore()
+const visualizationStore = useVisualizationStore()
+
+const expressionInfo = computed(
+  () => props.expressionId && graphStore.db.getExpressionInfo(props.expressionId),
+)
+const typeName = computed(() => expressionInfo.value?.typename ?? 'Any')
+
+/** The Visualization ID for the method to get the default visualization. */
+const defaultVisualizationVisualizationId = random.uuidv4() as Uuid
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/GraphVisualization.vue]: Why double `Visualization`? I think single is enough (it's obvious it's visualization ID... at least for me).

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue [hunk 3]
```diff
@@ -36,6 +39,62 @@ const emit = defineEmits<{
...
+          name: parsed.name,
+          module:
+            parsed.library == null
+              ? { kind: 'Builtin' }
+              : { kind: 'Library', name: parsed.library.name },
+        },
+      )
+    }
+    dataServer.on(`${OutboundPayload.VISUALIZATION_UPDATE}`, onVisualizationUpdate)
+  })
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/GraphVisualization.vue]: You may copy `@/util/observable.ts` [from my PR](https://github.com/enso-org/enso/pull/8385) and use `nextEvent` here.

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue [hunk 4]
```diff
@@ -36,6 +39,62 @@ const emit = defineEmits<{
...
+  ls.executeExpression(
+    projectStore.executionContext.id,
+    defaultVisualizationVisualizationId,
+    props.nodeId as ExprId,
+    `${ident}.default_visualization.to_js_object.to_json`,
+  )
+  return new Promise<VisualizationIdentifier | undefined>((resolve) => {
+    const onVisualizationUpdate = (payload: VisualizationUpdate, uuid: Uuid | null) => {
+      if (uuid !== defaultVisualizationVisualizationId) return
+      dataServer.off(`${OutboundPayload.VISUALIZATION_UPDATE}`, onVisualizationUpdate)
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/GraphVisualization.vue]: Instead of calling `off` inside, you may just use `once` instead of `on`.

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue [hunk 5]
```diff
@@ -36,6 +39,62 @@ const emit = defineEmits<{
...
+  const ident = props.pattern?.repr()
+  if (!ident) return
+  const ls = await projectStore.lsRpcConnection
+  const dataServer = await projectStore.dataConnection
+  ls.executeExpression(
+    projectStore.executionContext.id,
+    defaultVisualizationVisualizationId,
+    props.nodeId as ExprId,
+    `${ident}.default_visualization.to_js_object.to_json`,
+  )
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/GraphVisualization.vue]: Node's pattern may be many things. If it's `Point x y`, the expression would be `Point x y..default_visualization.to_js_object.to_json` what may give quite unexpected result.

And I think we should attach a "normal" visualization - this way we will just get the node's value without need of retrieving the node's pattern. And we could then react to value changes, for example when the Table stops having `longitude` and `latitude` columns, the default vis could change to `Table`.

**Review by farmaazon — CHANGES_REQUESTED** [review]

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue (hunk 2 above)
Comment (somebody1234) [reply app/gui2/src/components/GraphEditor/GraphVisualization.vue]: my thought process was, this is the visualization id to get the default visualization - not some kind of fallback visualization id for the *actual* visualization... but yeah it is kinda long

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue (hunk 4 above)
Comment (somebody1234) [reply app/gui2/src/components/GraphEditor/GraphVisualization.vue]: i can't - i have to check that the visualization ids are the same

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue (hunk 5 above)
Comment (somebody1234) [reply app/gui2/src/components/GraphEditor/GraphVisualization.vue]: that makes sense. actually i tried using a normal visualization before, but i'm not sure how to do dynamic dispatch (`.default_visualization.to_js_object.to_json`) on a visualization. i'll get back to you when i have some errors to show

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue (hunk 4 above)
Comment (farmaazon) [reply app/gui2/src/components/GraphEditor/GraphVisualization.vue]: Ah, indeed. Then, ignore also my previous comment.

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue (hunk 3 above)
Comment (farmaazon) [reply app/gui2/src/components/GraphEditor/GraphVisualization.vue]: Not applicable: see next comment

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue (hunk 5 above)
Comment (farmaazon) [reply app/gui2/src/components/GraphEditor/GraphVisualization.vue]: I think the preprocessor should be just `a -> a.default_visualization.to_js_object.to_json`

If this does not work, then we need a new method in `Visualizations` module in enso stdlib.

File: app/gui2/shared/yjsModel.ts (hunk 1 above)
Comment (somebody1234) [reply app/gui2/shared/yjsModel.ts]: i'll probably go with `visIdentifierEquals` accepting `null`s - `shared/` does not have access to `util`.

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue (hunk 5 above)
Comment (somebody1234) [reply app/gui2/src/components/GraphEditor/GraphVisualization.vue]: hmm... i wonder if `(${ident}).default_visualization.to_js_object.to_json` would be sufficient. but attaching a normal visualization is definitely better

**Comment by somebody1234** [comment]
spent way too long debugging an issue caused by me, when `return` (implicit `undefined` return) when converting to file format when `identifier` was `null`, meaning viz visibility was not saved when explicitly using `default_visualization`.

i haven't checked GUI1 but am assuming this must break compatibility with it, as the original impl lacked nullable `project` in the file storage format for viz metadata.

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue [hunk 6]
```diff
@@ -51,41 +51,29 @@ const expressionInfo = computed(
...
-              : { kind: 'Library', name: parsed.library.name },
-        },
-      )
+const configForGettingDefaultVisualization = computed<NodeVisualizationConfiguration | undefined>(
+  () => {
+    if (props.currentType) return
+    if (!props.expressionId) return
+    return {
+      visualizationModule: 'Standard.Visualization.Helpers',
+      expression: 'a -> a.default_visualization.to_js_object.to_json',
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/GraphVisualization.vue:68]: This should also work, but I leave it up to you.
```suggestion
      expression: '_.default_visualization.to_js_object.to_json',
```

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue [hunk 7]
```diff
@@ -51,41 +51,29 @@ const expressionInfo = computed(
...
+      expressionId: props.expressionId,
+    }
+  },
+)
+const defaultVisualizationRaw = projectStore.useVisualizationData(
+  configForGettingDefaultVisualization,
+) as ShallowRef<{ library: { name: string } | null; name: string } | undefined>
+const defaultVisualization = computed<VisualizationIdentifier | undefined>(() => {
+  const raw = defaultVisualizationRaw.value
+  return (
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/GraphVisualization.vue]: Why are these parentheses needed?

File: app/gui2/src/stores/graph/index.ts [hunk 8]
```diff
@@ -262,24 +262,11 @@ export const useGraphStore = defineStore('graph', () => {
 
   function normalizeVisMetadata(
     id: Opt<VisualizationIdentifier>,
-    visible?: boolean,
```
Comment (farmaazon) [inline app/gui2/src/stores/graph/index.ts:266]: Why this change? I always thought this is identical?

**Review by farmaazon — APPROVED** [review]

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue (hunk 7 above)
Comment (somebody1234) [reply app/gui2/src/components/GraphEditor/GraphVisualization.vue]: inserted by prettier. i can avoid them by reformatting to `if (!raw) return`

File: app/gui2/src/stores/graph/index.ts (hunk 8 above)
Comment (somebody1234) [reply app/gui2/src/stores/graph/index.ts:266]: this is (hopefully?) to ensure that users cannot omit `visible`, so we don't accidentally lose `visible` by forgettting to pass it

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue (hunk 7 above)
Comment (farmaazon) [reply app/gui2/src/components/GraphEditor/GraphVisualization.vue]: For me, it would be nicer to read.

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue (hunk 6 above)
Comment (somebody1234) [reply app/gui2/src/components/GraphEditor/GraphVisualization.vue:68]: i do prefer that, but i tested just now and... unfortunately that seems to not work - with a `visualizationEvaluationFailed` error saying "Message not supported."

File: app/gui2/src/stores/graph/index.ts (hunk 8 above)
Comment (farmaazon) [reply app/gui2/src/stores/graph/index.ts:266]: You're right, TS complains about missing argument.

File: app/gui2/src/stores/graph/index.ts [hunk 9]
```diff
@@ -9,7 +9,9 @@ import {
 } from '@/stores/graph/imports'
 import { useProjectStore } from '@/stores/project'
 import { useSuggestionDbStore } from '@/stores/suggestionDatabase'
-import { DEFAULT_VISUALIZATION_IDENTIFIER } from '@/stores/visualization'
+// FIXME: For some reason this is required, otherwise `compilerMessaging.ts` erros with:
+// Cannot access 'VisualizationContainer' before initialization
+import '@/stores/visualization'
```
Comment (Frizi) [inline app/gui2/src/stores/graph/index.ts]: There is a possible dependency cycle:
`@/components/VisualizationContainer.vue` <- `@/providers/visualizationConfig` <- `@/stores/visualization/compilerMessaging` <- `@/components/VisualizationContainer.vue`

Some of the imports in this chain are `import type`, but it seems like those still matter? If it does, then the first loaded module in this chain will determine which modules see which other ones fully defined during synchronous execution. You could try moving the `URLString` type into separate (util?) module, as it is the most common import and that would break the chain.

If this doesn't work, let's try to debug what's going on here. I'm worried that it can behave differently in production build, and cause the electron package to unexpectedly fail to load again.

File: app/gui2/ydoc-server/fileFormat.ts [hunk 10]
```diff
@@ -14,8 +14,8 @@ export type VisualizationMetadata = z.infer<typeof visualizationMetadata>
 const visualizationMetadata = z
   .object({
     show: z.boolean().default(true),
-    project: visualizationProject,
-    name: z.string(),
+    project: visualizationProject.nullable(),
```
Comment (Frizi) [inline app/gui2/ydoc-server/fileFormat.ts]: Is it really `nullable`, or rather `nullish`/`optional`? Not sure we should be saving nulls in the file.

**Review by Frizi — CHANGES_REQUESTED** [review]

File: app/gui2/ydoc-server/fileFormat.ts (hunk 10 above)
Comment (somebody1234) [reply app/gui2/ydoc-server/fileFormat.ts]: ah. it's a new addition, so the format can be whatever. i assume GUI1 doesn't support having visibility without an explicitly set viz type?

File: app/gui2/src/stores/graph/index.ts (hunk 9 above)
Comment (somebody1234) [reply app/gui2/src/stores/graph/index.ts]: man that was a pain... the issue was `VizContainer` -> `VizSelector` -> `vizStore`. seems like it might be quite painful to fix too...

**Comment by somebody1234** [comment]
merged develop into this branch because GitHub says there were conflicts (probably due to the moved files), but merge-ort had no issues

**Review by Frizi — APPROVED** [review]
